windows/devices/uwb/UwbAppConfiguration.cxx

```diff
@@ -0,0 +1,92 @@
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <numeric>
+
+#include <windows/devices/uwb/UwbAppConfiguration.hxx>
+
+bool
+operator==(const UWB_APP_CONFIG_PARAM& lhs, const UWB_APP_CONFIG_PARAM& rhs) noexcept
+{
+    return (lhs.size == rhs.size) && (std::memcmp(&lhs, &rhs, lhs.size) == 0);
+}
+
+using namespace windows::devices;
+
+IUwbAppConfigurationParameter::IUwbAppConfigurationParameter(UWB_APP_CONFIG_PARAM_TYPE parameterType, std::span<const uint8_t> parameterValue) :
+    m_buffer(offsetof(UWB_APP_CONFIG_PARAM, paramValue[parameterValue.size()])),
+    m_parameter(*reinterpret_cast<UWB_APP_CONFIG_PARAM*>(m_buffer.data()))
+{
+    m_parameter.size = std::size(m_buffer);
+    m_parameter.paramType = parameterType;
+    m_parameter.paramLength = parameterValue.size();
+    std::memcpy(&m_parameter.paramValue, parameterValue.data(), parameterValue.size());
+}
+
+std::size_t
+IUwbAppConfigurationParameter::DdiSize() const noexcept
+{
+    return std::size(m_buffer);
+}
+
+UWB_APP_CONFIG_PARAM&
+IUwbAppConfigurationParameter::DdiParameter() noexcept
+{
+    return m_parameter;
+}
+
+std::vector<uint8_t>&
+IUwbAppConfigurationParameter::DdiBuffer() noexcept
+{
+    return m_buffer;
+}
+
+namespace detail
+{
+/**
+ * @brief Calculate the total size in bytes of all the UWB_APP_CONFIG_PARAM DDI
+ * structures contained in the provided IUwbAppConfigurationParameter vector.
+ *
+ * @param parameters
+ * @return std::size_t
+ */
+std::size_t
+CalculateTotalUwbAppConfigurationBufferSize(const std::vector<std::shared_ptr<IUwbAppConfigurationParameter>>& parameters)
+{
+    return std::accumulate(parameters.cbegin(), parameters.cend(), static_cast<std::size_t>(0), [&](std::size_t totalSize, const auto& uwbAppConfigurationParameter) {
+        return totalSize + uwbAppConfigurationParameter->DdiSize();
+    });
+}
+} // namespace detail
+
+UwbSetAppConfigurationParameters::UwbSetAppConfigurationParameters(const std::vector<std::shared_ptr<IUwbAppConfigurationParameter>>& parameters, uint32_t sessionId) :
+    m_buffer(offsetof(UWB_APP_CONFIG_PARAMS, appConfigParams[0]) + ::detail::CalculateTotalUwbAppConfigurationBufferSize(parameters)), // TOOD: verify this is correct
+    m_parameters(*reinterpret_cast<UWB_SET_APP_CONFIG_PARAMS*>(std::data(m_buffer)))
+{
+    auto it = std::next(std::begin(m_buffer), offsetof(UWB_APP_CONFIG_PARAMS, appConfigParams[0]));
+
+    // Copy each complete UWB_APP_CONFIG_PARAM structure into the byte buffer.
+    for (const auto& parameter : parameters) {
+        const auto& parameterBuffer = parameter->DdiBuffer();
+        m_buffer.insert(it, std::cbegin(parameterBuffer), std::cend(parameterBuffer));
+        std::advance(it, std::size(parameterBuffer));
```

Comment:
I think we should use a simple byte array for m_buffer instead of a vector. Here, we already know exactly which parameters go in this struct, so there is no need for a dynamic vector. In addition, m_buffer.insert invalidates all iterators if more memory has been allocated due to its operation.

Reply:
The use of `std::vector` has many benefits over a simple byte array: use with other STL algorithms, re-sizing if necessary, easy copying to other containers, the list goes on.

Just because we don't need the grow the buffer in this instance, doesn't mean we shouldn't make use of the other features `std:;vector` has to offer. I might argue the use of `std::vector` here is worth it just for the fact that it contains both the data and the size together in one object. With a simple byte-buffer, we don't have that, and the state (data contents, size) could get out of sync.

Regarding `insert`: the vector was constructed with the exact number of bytes needed, so no re-allocation or iterator invalidations occur here.

Reply:
I'm fairly certain that insert always changes the size of vector. When I was debugging that was certainly the case. If we really want to use vector then we can't use insert here, but instead a for-each statement or something

Reply:
Also, the byte array is intended to hold a structure that does have an internal size field, so even though it's not pretty I think the benefit of the vector object also holding the size is not that big.

Reply:
> I'm fairly certain that insert always changes the size of vector. When I was debugging that was certainly the case. If we really want to use vector then we can't use insert here, but instead a for-each statement or something

Sorry, you're absolutely right about this. I mixed up `vector::insert` and `std::inserter`, the latter of which creates an `std::insert_iterator` which is used to insert at the specified iterator position. As you mentioned, the fix for the insertion is a range-based for works or `std::transform`.

> Also, the byte array is intended to hold a structure that does have an internal size field, so even though it's not pretty I think the benefit of the vector object also holding the size is not that big. Regarding the out-of-sync of the (data contents, size) state, that only happens right when it is initialized, and never again afterwards (unless the user deliberately overwrites the size field of the byte array).

These are all good points. Even so, is there any reason not to use `std::vector`; the benefit is clearly non-zero, so why not take that as win and go with it? `std::vector` is better then `std::unique_ptr<uint8_t[]>` in every case and scenario I can imagine, and I can't think of any place where `unique_ptr` would be preferred.

Reply:
I agree with the statement about no real drawback to using vector, and I switched back to using vector